listener/listener.go

```diff
@@ -300,80 +315,150 @@ func (l *Listener) Stream(ctx context.Context) error {
-				l.monitor.IncProblematicEvents(problemKindPublish)
-				return fmt.Errorf("publish: %w", err)
+			if msg == nil {
+				continue
 			}
 
-			l.monitor.IncPublishedEvents(subjectName, event.Table)
+			messageChan <- msg
+		}
+	})
 
-			l.log.Info(
-				"event was sent",
-				slog.String("subject", subjectName),
-				slog.String("action", event.Action),
-				slog.String("table", event.Table),
-				slog.Uint64("lsn", l.readLSN()),
-			)
+	group.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				l.log.Warn("stream: context canceled", "err", ctx.Err())
+				return nil
+			case msg := <-messageChan:
+				if err := l.parser.ParseWalMessage(msg.WalMessage.WalData, tx); err != nil {
+					l.monitor.IncProblematicEvents(problemKindParse)
+					return fmt.Errorf("parse: %w", err)
+				}
+
+				if tx.CommitTime != nil {
+					events := tx.CreateEventsWithFilter(ctx, l.cfg.Listener.Filter.Tables)
+
+					eventsChan <- &messageAndEvents{
+						msg,
+						events,
+					}
+					tx.Clear()
+				}
+
+				l.processHeartBeat(msg)
+			}
+		}
+	})
 
-			tx.pool.Put(event)
+	group.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				l.log.Warn("stream: context canceled", "err", ctx.Err())
+				return nil
+			case object := <-eventsChan:
+				events, msg := object.events, object.msg
+
+				for idx, event := range events {
+					subjectName := event.SubjectName(l.cfg)
+					topicSet[subjectName] = true
+
+					result := l.publisher.Publish(ctx, subjectName, event)
+
+					// Include the message on the last event of the batch
+					if idx == len(events)-1 {
+						resultChan <- &eventAndPublishResult{
+							subjectName: subjectName,
+							event:       event,
+							result:      result,
+							msg:         msg,
+						}
+					} else {
+						resultChan <- &eventAndPublishResult{
+							subjectName: subjectName,
+							event:       event,
+							result:      result,
+						}
+					}
+				}
+			}
 		}
+	})
 
-		tx.Clear()
-	}
+	group.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				l.log.Warn("stream: context canceled", "err", ctx.Err())
+				return nil
+			case object := <-resultChan:
+				subjectName, result, event, msg := object.subjectName, object.result, object.event, object.msg
+
+				_, err := result.Get(ctx)
+				if err != nil {
+					l.monitor.IncProblematicEvents(problemKindPublish)
+					// FIXME: don't tear down the streamer when an error fails to be published
+					return fmt.Errorf("publish: %w", err)
+				}
+
+				l.monitor.IncPublishedEvents(subjectName, event.Table)
+
+				l.log.Info(
+					"event was sent",
+					slog.String("subject", subjectName),
+					slog.String("action", event.Action),
+					slog.String("table", event.Table),
+					slog.Uint64("lsn", l.readLSN()),
+				)
+
+				tx.pool.Put(event)
+
+				if msg != nil {
+					if msg.WalMessage.WalStart > l.readLSN() {
+						if err := l.AckWalMessage(msg.WalMessage.WalStart); err != nil {
```

Comment:
One small downside to the message at a time approach over this channel is that we ack every message, instead of every batch. I don't think its a huge deal, but I think secretly lying beneath this call is a whole roundtrip to postgres that doesn't need to happen every single time, and because we're at-least-once it seems fine to defer it sometimes.

Reply:
Every message from the WAL turns into a batch of events, and this pipeline will only ACK once per message (so once per batch of events).

This is similar to how it's done upstream, your implementation would call this check once per event if I'm reading it correctly.

Reply:
> this pipeline will only ACK once per message (so once per batch of events).

Cause of the numeric condition?

Reply:
Ah, I see! That's a little implicit but works for me!